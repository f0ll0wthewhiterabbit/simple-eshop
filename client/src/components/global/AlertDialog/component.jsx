import React from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'

const AlertDialog = ({
  isModalOpened,
  title,
  children,
  closeModal,
  storeFieldName,
  deleteItems,
}) => {
  const handleConfirmButton = () => {
    deleteItems(storeFieldName)
  }

  return (
    <Dialog
      open={isModalOpened}
      onClose={closeModal}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
    >
      <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
      {children && (
        <DialogContent>
          <DialogContentText id="alert-dialog-description">{children}</DialogContentText>
        </DialogContent>
      )}
      <DialogActions>
        <Button onClick={closeModal} color="primary">
          Cancel
        </Button>
        <Button onClick={handleConfirmButton} color="primary" autoFocus>
          Confirm
        </Button>
      </DialogActions>
    </Dialog>
  )
}

AlertDialog.propTypes = {
  isModalOpened: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  closeModal: PropTypes.func.isRequired,
  storeFieldName: PropTypes.string,
  deleteItems: PropTypes.func.isRequired,
}

export default AlertDialog
