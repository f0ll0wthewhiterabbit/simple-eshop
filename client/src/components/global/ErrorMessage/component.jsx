import React from 'react'
import PropTypes from 'prop-types'
import PriorityHighOutlinedIcon from '@material-ui/icons/PriorityHighOutlined'

import { Wrapper, IconWrapper, Heading, Message } from './styles'

const ErrorMessage = ({ children, title }) => {
  return (
    <Wrapper>
      <IconWrapper>
        <PriorityHighOutlinedIcon />
      </IconWrapper>
      {title && (
        <Heading variant="h5" component="h1">
          {title}
        </Heading>
      )}
      {children && <Message variant="h6">{children}</Message>}
    </Wrapper>
  )
}

ErrorMessage.defaultProps = {
  children: '',
  title: '',
}

ErrorMessage.propTypes = {
  children: PropTypes.string,
  title: PropTypes.string,
}

export default ErrorMessage
