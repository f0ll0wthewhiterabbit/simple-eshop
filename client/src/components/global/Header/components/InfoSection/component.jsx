import React from 'react'

import {
  Wrapper,
  InfoList,
  InfoItem,
  InfoText,
  ShippingIcon,
  StudentIcon,
  MoneyIcon,
} from './styles'

const InfoSection = () => {
  return (
    <Wrapper>
      <InfoList>
        <InfoItem>
          <ShippingIcon />
          <InfoText>Free shipping on orders over $30</InfoText>
        </InfoItem>
        <InfoItem>
          <StudentIcon />
          <InfoText>20% Student Discount</InfoText>
        </InfoItem>
        <InfoItem>
          <MoneyIcon />
          <InfoText>30% off on dresses. Use code: 30OFF</InfoText>
        </InfoItem>
      </InfoList>
    </Wrapper>
  )
}

export default InfoSection
