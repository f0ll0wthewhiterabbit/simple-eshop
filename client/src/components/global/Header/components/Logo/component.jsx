import React from 'react'
import PropTypes from 'prop-types'

import { HOME_PAGE_PATH } from '../../../../../constants'
import { Wrapper, SiteLink, LogoImg } from './styles'

const Logo = ({ isAdmin, isDarkTheme }) => {
  return (
    <Wrapper justify={isAdmin ? 'center' : 'flex-start'}>
      <SiteLink to={HOME_PAGE_PATH}>
        <LogoImg src="/logo.png" alt="Logo" isDarkTheme={isDarkTheme} />
      </SiteLink>
    </Wrapper>
  )
}

Logo.propTypes = {
  isAdmin: PropTypes.bool.isRequired,
  isDarkTheme: PropTypes.bool.isRequired,
}

export default Logo
