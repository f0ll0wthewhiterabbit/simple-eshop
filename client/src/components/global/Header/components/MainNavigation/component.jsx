import React from 'react'

import { Wrapper, NavigationList, NavigationItem, NavigationLink } from './styles'

const MainNavigation = () => {
  return (
    <Wrapper>
      <NavigationList>
        <NavigationItem>
          <NavigationLink href="!#">Home</NavigationLink>
        </NavigationItem>
        <NavigationItem>
          <NavigationLink href="!#">Shop</NavigationLink>
        </NavigationItem>
        <NavigationItem>
          <NavigationLink href="!#">About</NavigationLink>
        </NavigationItem>
        <NavigationItem>
          <NavigationLink href="!#">Blog</NavigationLink>
        </NavigationItem>
        <NavigationItem>
          <NavigationLink href="!#">Contact</NavigationLink>
        </NavigationItem>
      </NavigationList>
    </Wrapper>
  )
}

export default MainNavigation
