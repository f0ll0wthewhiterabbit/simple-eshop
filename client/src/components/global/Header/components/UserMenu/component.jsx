import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import { withTheme } from 'styled-components'
import { Menu, MenuItem, ListItemText, useMediaQuery } from '@material-ui/core'
import HighlightOffOutlinedIcon from '@material-ui/icons/HighlightOffOutlined'
import ExitToAppOutlinedIcon from '@material-ui/icons/ExitToAppOutlined'
import PersonOutlineOutlinedIcon from '@material-ui/icons/PersonOutlineOutlined'
import AccountBoxIcon from '@material-ui/icons/AccountBox'

import {
  Wrapper,
  UserMenuButtonNormal,
  UserMenuButtonSmall,
  ArrowIcon,
  IconWrapper,
} from './styles'
import { STORE_FIELD_CURRENT_USER } from '../../../../../constants'

const UserMenu = ({
  theme,
  userName,
  isAdmin,
  isDeleteRequestSent,
  showModal,
  signOut,
  history,
  location,
}) => {
  const [anchorEl, setAnchorEl] = React.useState(null)

  const isMobileDevice = !useMediaQuery(theme.breakpoints.up('sm'))

  const handleClick = event => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  const handleDeleteAccountClick = () => {
    showModal(STORE_FIELD_CURRENT_USER)
    handleClose()
  }

  const handleSignOutClick = () => {
    signOut(history, location)
  }

  const userMenuButton = isMobileDevice ? (
    <UserMenuButtonSmall aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
      <AccountBoxIcon />
    </UserMenuButtonSmall>
  ) : (
    <UserMenuButtonNormal aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
      {userName} <ArrowIcon />
    </UserMenuButtonNormal>
  )

  return (
    <Wrapper>
      {userMenuButton}
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <MenuItem disabled>
          <IconWrapper>
            <PersonOutlineOutlinedIcon fontSize="small" />
          </IconWrapper>
          <ListItemText primary={userName} />
        </MenuItem>
        {!isAdmin && !isDeleteRequestSent && (
          <MenuItem onClick={handleDeleteAccountClick}>
            <IconWrapper>
              <HighlightOffOutlinedIcon fontSize="small" />
            </IconWrapper>
            <ListItemText primary="Delete Account" />
          </MenuItem>
        )}
        <MenuItem onClick={handleSignOutClick}>
          <IconWrapper>
            <ExitToAppOutlinedIcon fontSize="small" />
          </IconWrapper>
          <ListItemText primary="Sign Out" />
        </MenuItem>
      </Menu>
    </Wrapper>
  )
}

UserMenu.propTypes = {
  theme: PropTypes.shape({
    breakpoints: PropTypes.object.isRequired,
  }).isRequired,
  userName: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  isDeleteRequestSent: PropTypes.bool.isRequired,
  showModal: PropTypes.func.isRequired,
  signOut: PropTypes.func.isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
  location: PropTypes.shape({ pathname: PropTypes.string.isRequired }).isRequired,
}

export default withRouter(withTheme(UserMenu))
