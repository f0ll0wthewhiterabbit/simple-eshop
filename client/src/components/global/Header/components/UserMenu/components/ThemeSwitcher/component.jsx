import React from 'react'
import PropTypes from 'prop-types'
import { Switch } from '@material-ui/core'

import { FIELD_THEME_DEFAULT, FIELD_THEME_DARK } from '../../../../../../../constants'
import { Wrapper } from './styles'

const ThemeSwitcher = ({ isDarkTheme, toggleTheme }) => {
  const handleThemeChange = () => {
    const nextTheme = isDarkTheme ? FIELD_THEME_DEFAULT : FIELD_THEME_DARK
    toggleTheme(nextTheme)
  }

  return (
    <Wrapper>
      <Switch checked={isDarkTheme} onChange={handleThemeChange} size="small" />
    </Wrapper>
  )
}

ThemeSwitcher.propTypes = {
  isDarkTheme: PropTypes.bool.isRequired,
  toggleTheme: PropTypes.func.isRequired,
}

export default ThemeSwitcher
