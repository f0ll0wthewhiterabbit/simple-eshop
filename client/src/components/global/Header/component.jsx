import React from 'react'
import PropTypes from 'prop-types'
import { Toolbar, Container } from '@material-ui/core'
import MenuIcon from '@material-ui/icons/Menu'

import Logo from './components/Logo'
import UserMenu from './components/UserMenu'
import { Wrapper, MenuButton, WarningMessage } from './styles'

const Header = ({ openSidebar, isAdmin, isAuthenticated, isDeleteRequestSent }) => {
  const handleMenuButtonClick = () => {
    openSidebar()
  }

  return (
    <Wrapper position="static">
      <Container maxWidth="lg">
        <Toolbar>
          {isAdmin && (
            <MenuButton color="secondary" aria-label="open sidebar" onClick={handleMenuButtonClick}>
              <MenuIcon />
            </MenuButton>
          )}
          <Logo />
          {isAuthenticated && isDeleteRequestSent && (
            <WarningMessage variant="body2" color="error">
              Unfortunately, your account will be deleted soon.
            </WarningMessage>
          )}
          {isAuthenticated && <UserMenu />}
        </Toolbar>
      </Container>
    </Wrapper>
  )
}

Header.propTypes = {
  openSidebar: PropTypes.func.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  isDeleteRequestSent: PropTypes.bool.isRequired,
}

export default Header
