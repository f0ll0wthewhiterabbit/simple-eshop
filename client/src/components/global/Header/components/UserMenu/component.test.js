import React from 'react'
import { shallow } from 'enzyme'
import useMediaQuery from '@material-ui/core/useMediaQuery'

import { UserMenu } from './component'
import findByTestAttr from '../../../../../utils/findByTestAttr'

jest.mock('@material-ui/core/useMediaQuery')

describe('UserMenu component', () => {
  let wrapper
  const initialProps = {
    theme: { breakpoints: { up: jest.fn() } },
    userName: 'testUserName',
    isAdmin: false,
    isDeleteRequestSent: false,
    showModal: jest.fn(),
    signOutRequest: jest.fn(),
    history: { push: jest.fn() },
    location: { pathname: '/testPathName' },
  }
  const dataTestMenu = 'menu'
  const dataTestDeleteAccountButton = 'deleteAccountButton'
  const dataTestMenuButton = 'menuButton'
  const dataTestProfileButton = 'profileButton'
  const dataTestSignOutButton = 'signOutButton'

  const setState = jest.fn()
  const useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation(init => [init, setState])

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<UserMenu {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  afterEach(() => {
    initialProps.showModal.mockClear()
    initialProps.signOutRequest.mockClear()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should contain delete account button', () => {
    const button = findByTestAttr(wrapper, dataTestDeleteAccountButton)

    expect(button).toHaveLength(1)
  })

  it(`shouldn't contain delete account button if delete account request is sent`, () => {
    wrapper = generateWrapper({ isDeleteRequestSent: true })
    const button = findByTestAttr(wrapper, dataTestDeleteAccountButton)

    expect(button).toHaveLength(0)
  })

  it(`shouldn't contain delete account button in admin mode`, () => {
    wrapper = generateWrapper({ isAdmin: true })
    const button = findByTestAttr(wrapper, dataTestDeleteAccountButton)

    expect(button).toHaveLength(0)
  })

  it('should render button with user name on medium and large desktop width', () => {
    useMediaQuery.mockImplementation(() => true)
    wrapper = generateWrapper()
    const button = findByTestAttr(wrapper, dataTestMenuButton)

    expect(button.text().indexOf(initialProps.userName)).not.toBe(-1)
  })

  it('should render button without user name on small desktop width', () => {
    useMediaQuery.mockImplementation(() => false)
    wrapper = generateWrapper()
    const button = findByTestAttr(wrapper, dataTestMenuButton)

    expect(button.text().indexOf(initialProps.userName)).toBe(-1)
  })

  describe('closed menu', () => {
    it(`should initially render with closed menu`, () => {
      const menu = findByTestAttr(wrapper, dataTestMenu)

      expect(menu.props().open).toBe(false)
    })

    it('should open menu after button click', () => {
      const button = findByTestAttr(wrapper, dataTestMenuButton)
      const target = 'testTarget'
      button.simulate('click', { currentTarget: target })

      expect(setState).toHaveBeenCalledWith(target)
    })
  })

  describe('opened menu', () => {
    it(`should show modal after delete account button click`, () => {
      const deleteAccountButton = findByTestAttr(wrapper, dataTestDeleteAccountButton)
      deleteAccountButton.simulate('click')

      expect(initialProps.showModal).toHaveBeenCalledTimes(1)
    })

    it(`should close menu after delete account button click`, () => {
      const menu = findByTestAttr(wrapper, dataTestMenu)
      const deleteAccountButton = findByTestAttr(wrapper, dataTestDeleteAccountButton)
      deleteAccountButton.simulate('click')

      expect(menu.props().open).toBe(false)
    })

    it(`should close menu after profile button click`, () => {
      const menu = findByTestAttr(wrapper, dataTestMenu)
      const profileButton = findByTestAttr(wrapper, dataTestProfileButton)
      profileButton.simulate('click')

      expect(menu.props().open).toBe(false)
    })

    it(`should sign out after sign out button click`, () => {
      const signOutButton = findByTestAttr(wrapper, dataTestSignOutButton)
      signOutButton.simulate('click')

      expect(initialProps.signOutRequest).toHaveBeenCalledTimes(1)
    })
  })
})
