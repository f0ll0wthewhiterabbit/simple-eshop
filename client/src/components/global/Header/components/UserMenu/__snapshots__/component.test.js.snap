// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu component should render correctly 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(IconButton)))
    aria-controls="simple-menu"
    aria-haspopup="true"
    data-test="menuButton"
    onClick={[Function]}
  >
    <AccountBoxIcon />
  </Styled(WithStyles(ForwardRef(IconButton)))>
  <Styled(WithStyles(ForwardRef(Menu)))
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    data-test="menu"
    getContentAnchorEl={null}
    id="simple-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <WithStyles(ForwardRef(MenuItem))
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Link",
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      data-test="profileButton"
      onClick={[Function]}
      to="/profile"
    >
      <Styled(WithStyles(ForwardRef(ListItemIcon)))>
        <PersonOutlineOutlinedIcon
          fontSize="small"
        />
      </Styled(WithStyles(ForwardRef(ListItemIcon)))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Profile"
      />
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      data-test="deleteAccountButton"
      onClick={[Function]}
    >
      <Styled(WithStyles(ForwardRef(ListItemIcon)))>
        <HighlightOffOutlinedIcon
          fontSize="small"
        />
      </Styled(WithStyles(ForwardRef(ListItemIcon)))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Delete Account"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </Styled(WithStyles(ForwardRef(Menu)))>
  <Connect(ThemeSwitcher) />
  <Styled(WithStyles(ForwardRef(IconButton)))
    aria-label="sign-out"
    data-test="signOutButton"
    onClick={[Function]}
  >
    <ExitToAppIcon />
  </Styled(WithStyles(ForwardRef(IconButton)))>
</styled.div>
`;
